### PR TITLE
docs(saml): session library v4.0.0

### PR DESCRIPTION
**Pull request by samugi**
### Description

Documentation for https://github.com/Kong/kong-ee/pull/4395 / [KAG-235](https://konghq.atlassian.net/browse/KAG-235)

### Checklist 

- [X] Review label added
- [X] PR pointed to correct branch

[KAG-235]: https://konghq.atlassian.net/browse/KAG-235?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/_hub/kong-inc/saml/_index.md [hunk 1]
```diff
@@ -216,16 +216,41 @@ params:
...
+        The session cookie rolling timeout in seconds.
+        Specifies how long the session can be used until it needs to be renewed.
+    - name: session_absolute_timeout
+      minimum_version: "3.2.x"
+      required: false
+      default: 86400
+      datatype: integer
+      description: |
+        The session cookie absolute timeout in seconds.
+        Specifies how long the session can be used until it is no longer valid
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:239]:
```suggestion
        Specifies how long the session can be used until it is no longer valid.
```

File: app/_hub/kong-inc/saml/_index.md [hunk 2]
```diff
@@ -250,7 +276,25 @@ params:
...
         - `None`: Cookies will be sent in all contexts, including responses to both first-party and cross-origin requests. If `SameSite=None` is set, the cookie secure attribute must also be set or the cookie will be blocked.
         - `off`: Do not set the Same-Site flag.
+    - name: session_cookie_same_site
+      minimum_version: "3.2.x"
+      required: false
+      default: Lax
+      datatype: string
+      description: |
+        Controls whether a cookie is sent with cross-origin requests, providing some protection against cross-site request forgery attacks:
+        - `Strict`: Cookies will only be sent in a first-party context and not be sent along with requests initiated by third party websites.
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:286]:
```suggestion
        - `Strict`: Cookies will only be sent in a first-party context and aren't sent along with requests initiated by third party websites.
```

File: app/_hub/kong-inc/saml/_index.md [hunk 3]
```diff
@@ -250,7 +276,25 @@ params:
...
+        - `Lax`: Cookies are not sent on normal cross-site subrequests, like loading images or frames into a third party site, but are sent when a user is navigating to the origin site, like when they are following a link.
+        - `None`: Cookies will be sent in all contexts, including responses to both first-party and cross-origin requests. If `SameSite=None` is set, the cookie secure attribute must also be set or the cookie will be blocked.
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:288]:
```suggestion
        - `None`: Cookies will be sent in all contexts, including responses to both first party and cross-origin requests. If `SameSite=None` is set, the cookie secure attribute must also be set or the cookie will be blocked.
```

File: app/_hub/kong-inc/saml/_index.md [hunk 4]
```diff
@@ -250,7 +276,25 @@ params:
...
+        - `Default`: Do not explicitly specify a Same-Site attribute.
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:289]: Wouldn't the default be `Lax`? Or is `Default` a separate option from `Lax`?

File: app/_hub/kong-inc/saml/_index.md [hunk 5]
```diff
@@ -288,21 +334,104 @@ params:
...
       description: |
         The session strategy:
         - `none`: no compression.
         - `zlib`: use Zlib to compress cookie data.
+    - name: session_audience
+      minimum_version: "3.2.x"
+      required: false
+      default: '"default"'
+      datatype: string
+      description: The session audience e.g. "my-application"
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:350]:
```suggestion
      description: The session audience, for example "my-application".
```

File: app/_hub/kong-inc/saml/_index.md [hunk 6]
```diff
@@ -288,21 +334,104 @@ params:
...
+    - name: session_request_headers
+      minimum_version: "3.2.x"
+      required: false
+      default: null
+      datatype: array of string elements
+      description: |
+        List of information to include (as headers) in the request to upstream. 
+        Accepted values are: `id`, `audience`, `subject`, `timeout`, `idling-timeout`, `rolling-timeout`,
+        `absolute-timeout`.
+        E.g. { "id", "timeout" } will set both `Session-Id` and `Session-Timeout` in the request headers.
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:384]:
```suggestion
        Accepted values are: `id`, `audience`, `subject`, `timeout`, `idling-timeout`, `rolling-timeout`, and
        `absolute-timeout`.
        For example, { "id", "timeout" } will set both `Session-Id` and `Session-Timeout` in the request headers.
```

File: app/_hub/kong-inc/saml/_index.md [hunk 7]
```diff
@@ -288,21 +334,104 @@ params:
...
+    - name: session_response_headers
+      minimum_version: "3.2.x"
+      required: false
+      default: null
+      datatype: array of string elements
+      description: |
+        List of information to include (as headers) in the response to downstream. 
+        Accepted values are: `id`, `audience`, `subject`, `timeout`, `idling-timeout`, `rolling-timeout`,
+        `absolute-timeout`.
+        E.g. { "id", "timeout" } will inject both `Session-Id` and `Session-Timeout` in the response headers.
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:394]:
```suggestion
        Accepted values are: `id`, `audience`, `subject`, `timeout`, `idling-timeout`, `rolling-timeout`, and
        `absolute-timeout`.
        For example: { "id", "timeout" } will inject both `Session-Id` and `Session-Timeout` in the response headers.
```

File: app/_hub/kong-inc/saml/_index.md [hunk 8]
```diff
@@ -288,21 +334,104 @@ params:
...
+    - name: session_store_metadata
+      minimum_version: "3.2.x"
+      required: false
+      default: false
+      datatype: boolean
+      description: | 
+        Configures whether or not session metadata should be stored.
+        This includes information about what are the active sessions for a specific audience
+        belonging to a specific subject.
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:403]: Would the "specific audience" and "specific subjects" be determined by the following params?

File: app/_hub/kong-inc/saml/_index.md [hunk 9]
```diff
@@ -288,21 +334,104 @@ params:
...
+      description: |
+        When set to `true`, the value of subject is hashed before being stored.
+        Only applies when `session_store_metadata` is enabled.
+    - name: session_hash_storage_key
+      minimum_version: "3.2.x"
+      required: false
+      default: false
+      datatype: boolean
+      description: |
+        When set to `true`, the storage key (session id) is hashed for extra security.
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:424]:
```suggestion
        When set to `true`, the storage key (session ID) is hashed for extra security.
```

File: app/_hub/kong-inc/saml/_index.md [hunk 10]
```diff
@@ -524,3 +688,41 @@ curl --request POST \
 2. The browser is redirected to the AzureAD Sign in page. Enter the user credentials of a user configured in AzureAD
 3. If user credentials are valid, the browser will be redirected to https://httpbin.org/anything
 4. If the user credentials are invalid, a `401` unauthorized HTTP status code is returned
+
+## Changelog
+
+**{{site.base_gateway}} 3.2.x**
+* The plugin has been updated to use version 4.0.0 of the `lua-resty-session` library
+which introduced several new features such as the possibility to specify audiences.
+The following configuration parameters have been affected:
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:697]:
```suggestion
* The plugin has been updated to use version 4.0.0 of the `lua-resty-session` library. This introduced several new features, such as the possibility to specify audiences.
The following configuration parameters were affected:
```

File: app/_hub/kong-inc/saml/_index.md [hunk 11]
```diff
@@ -524,3 +688,41 @@ curl --request POST \
...
+Renamed:
+  * From: `session_cookie_lifetime` to `session_rolling_timeout`
+  * From: `session_cookie_idletime` to `session_idling_timeout`
+  * From: `session_cookie_samesite` to `session_cookie_same_site`
+  * From: `session_cookie_httponly` to `session_cookie_http_only`
+  * From: `session_memcache_prefix` to `session_memcached_prefix`
+  * From: `session_memcache_socket` to `session_memcached_socket`
+  * From: `session_memcache_host` to `session_memcached_host`
+  * From: `session_memcache_port` to `session_memcached_port`
+  * From: `session_redis_cluster_maxredirections` to `session_redis_cluster_max_redirections`
```
Comment (cloudjumpercat) [inline app/_hub/kong-inc/saml/_index.md:722]:
```suggestion
  * `session_cookie_lifetime` to `session_rolling_timeout`
  * `session_cookie_idletime` to `session_idling_timeout`
  * `session_cookie_samesite` to `session_cookie_same_site`
  * `session_cookie_httponly` to `session_cookie_http_only`
  * `session_memcache_prefix` to `session_memcached_prefix`
  * `session_memcache_socket` to `session_memcached_socket`
  * `session_memcache_host` to `session_memcached_host`
  * `session_memcache_port` to `session_memcached_port`
  * `session_redis_cluster_maxredirections` to `session_redis_cluster_max_redirections`
```

**Review by cloudjumpercat — COMMENTED** [review]
Left a few suggestions and questions, thanks for updating this plugin doc!

File: app/_hub/kong-inc/saml/_index.md (hunk 4 above)
Comment (samugi) [reply app/_hub/kong-inc/saml/_index.md:289]: I know. Sounds weird, but `Default` is in fact a separate option which... is not the default :smile:

File: app/_hub/kong-inc/saml/_index.md (hunk 8 above)
Comment (samugi) [reply app/_hub/kong-inc/saml/_index.md:403]: only the audience is configurable, the subject is inferred automatically. The configuration for the audience is controlled by the parameter: `session_audience`

**Comment by samugi** [comment]
thanks for the great review @cloudjumpercat !

**Pull request merged**